tests/licensedcode/test_models.py

```
@@ -534,3 +537,68 @@ def test_Rule__validate_with_false_positive_rule(self):
         rule_dir = self.get_test_loc('models/rule_validate')
         rule = list(models.load_rules(rule_dir))[0]
         assert list(rule.validate()) == []
+
+    def test_key_phrases_yields_spans(self):
+        rule_stored_text = (
+            'This released software is {{released}} by under {{the MIT license}}. '
+            'Which is a license originating at Massachusetts Institute of Technology (MIT).'
+        )
+        rule = models.Rule(license_expression='mit', stored_text=rule_stored_text)
+
+        key_phrases = rule.key_phrases()
+
+        assert list(key_phrases) == [models.Span(4), models.Span(7, 9)]
+
+
+class TestGetKeyPhrases(TestCaseClass):
+    def test_get_key_phrases_yields_spans(self):
+        text = (
+            'This released software is {{released}} by under {{the MIT license}}. '
+            'Which is a license originating at Massachusetts Institute of Technology (MIT).'
+        )
+
+        key_phrases = get_key_phrases(text)
+
+        assert list(key_phrases) == [models.Span(4), models.Span(7, 9)]
+
+    def test_get_key_phrases_yields_nothing_when_unclosed_key_phrase_markup(self):
```

IMHO we should fail loudly if there is any key phrase markup that is incomplete and not well balanced. The principle here would be to always ensure that the rule and license data is correct and that any incorrect data should fail tests.

---

I've updated the code to do so and I am not running the test suite and found that some issues will not pass when being so strict. For example `src/licensedcode/data/rules/lgpl-3.0-plus_27.RULE` contains the following:

```
GNU Lesser General Public License, Version 3 {{{ */

 is free software: you can redistribute it and/or modify it under
the terms of the GNU Lesser General Public License as published by the
Free Software Foundation, either version 3 of the License, or (at your
option) any later version.

 is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
License for more details.

You should have received a copy of the GNU Lesser General Public License
along with .  If not, see <http://www.gnu.org/licenses/>.
```

Ways to work around it:

* Remove `{{{` from the rule, it wouldn't have any effect because it would be filtered out anyway?
* Come up with an alternative syntax that doesn't conflict.
  * Likely to become a problem in the future again, unless choosing something odd enough.
* Mark rule as having key phrases, but this feels cumbersome.

---

@petergardfjall Do you have an opinion on this?